File: day2/python/part2/day2_2.py

import sys

def main():
    """
    AOC 2024 Day 2 Part 2: Reads each level of input and checks if they are safe.
    If not safe, checks if removing a single element makes the level safe. Counts how many levels
    are safe based on the following onditions:

    - Sequence must either be strictly increasing or strictly decreasing.
    - The difference between adjacent values must be between 1 and 3.
    - If the level is unsafe, attempts to remove one element at a time and check if the resulting
      sequence is safe.

    Outputs the count of safe levels.
    """
    safe = 0  # Counter for safe levels

    while True:
        try:
            # Read a line of input, strip leading/trailing whitespace
            line = input().strip()
            if not line:
                break  # break if empty

            # parse the line into a list of integers
            level = list(map(int, line.split()))

            # check if safe, else attempt to remove
            if check_safe(level):
                safe += 1
            elif check_removal(level):
                safe += 1

        except EOFError:
            break  # end loop on EOF

    # output count
    print(f"# safe levels: {safe}")


def check_safe(level):
    """
    Checks if the given level is safe based on the following conditions:
    - Sequence must either be strictly increasing or strictly decreasing.
    - The difference between adjacent values must be between 1 and 3.
    
    Args:
    - level (list of int): List of integers representing the report level.

    Returns:
    - bool: True if the level is safe, False otherwise.
    """
    dec = False  # track if the sequence is decreasing
    inc = False  # track if the sequence is increasing

    for i in range(1, len(level)):
        # Check if incrementing or decrementing, unsafe if neither
        if level[i] > level[i - 1]:
            inc = True
        elif level[i] < level[i - 1]:
            dec = True
        else:
            return False  # Sequence is neither increasing nor decreasing

        # Check if the difference is within the threshold of 1 to 3
        diff = abs(level[i] - level[i - 1])
        if diff > 3 or diff < 1:
            return False

        # Unsafe if both increasing and decreasing trends are found
        if dec and inc:
            return False

    return True


def check_removal(vec):
    """
    Tries removing a single element from the vector and checks if the resulting level is safe.
    
    Args:
    - vec (list of int): List of integers representing the report level.

    Returns:
    - int: 1 if the level is safe after removal of one element, 0 otherwise.
    """
    for idx in range(len(vec)):
        # Create a temporary vector and remove the element at index 'idx'
        temp = vec[:idx] + vec[idx+1:]

        # Check if the level is safe after removal
        if check_safe(temp):
            return 1  # The level is safe after removal

    return 0  # The level is unsafe even after removal


if __name__ == "__main__":
    main()

File: day2/c++/part2/day2_2.cpp
/**
 * @file day2_2.cpp
 * @author ProjectZuki
 * @brief AOC 2024 Day 2 Part 2: Determines safety levels based on defined constraints.
 *          Removing one element at at time if unsafe, to check if still safe.
 * @date 2024-12-02
 * 
 */

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

istream& operator>>(istream& is, vector<int>& vec);
bool checkSafe(vector<int> level);
int checkRemoval(vector<int> vec);

/**
 * @brief Main entry point for the program. Reads each level of input, checking
 *          if they are safe. If not safe, check to see if a single element
 *          can be removed to make it safe.
 * 
 * @return EXIT_SUCCESS if success
 */
int main() {

    int safe = 0;

    vector<int> level;
    while (cin >> level) {
        safe += checkSafe(level)? 1 : checkRemoval(level);
    }

    cout << "# safe levels: " << safe << endl;

    return EXIT_SUCCESS;
}

/**
 * @brief Overload ('>>') to read one line of whitespace-separated
 *          integers into the vector
 * 
 * @param is input stream
 * @param vec vector to store values
 * @return istream&  Ref to input stream
 */
istream& operator>>(istream& is, vector<int>& vec) {
    vec.clear();
    int num;

    while (is.peek() != '\n' && is.peek() != EOF && is >> num) {
        vec.push_back(num);
    }
    is.ignore();
    
    return is;
}

/**
 * @brief checks if the current level is safe based on conditions:
 *              - Sequence must increase OR decrease
 *              - Difference must be between 1 - 3
 * 
 * @param level vector of report values
 * @return true if safe, else false
 */
bool checkSafe(vector<int> level) {
    bool dec = false;
    bool inc = false;

    for (size_t i = 1; i < level.size(); i++) {

        // check if incrementing or decrementing, unsafe if neither
        if (level.at(i) > level.at(i-1)) {
            inc = true;
        } else if (level.at(i) < level.at(i-1)) {
            dec = true;
        } else {
            return false;
        }

        // check threshold of inc/dec
        auto diff = abs(level.at(i) - level.at(i-1));
        if (diff > 3 || diff < 1) {
            return false;
        }

        // unsafe if incrementing and decrementing
        if (dec && inc) {
            return false;
        } 
    }

    return true;
}


/**
 * @brief Removes a single element at a time and checks if the level is still valid.
 * 
 * @param vec vector of integers representing the report level
 * @return 1 if safe, else 0
 */
int checkRemoval(vector<int> vec) {

    int safe = 0;               // 1 if safe, else 0
    vector<int> temp = vec;     // temp vector to erase elements with

    for (size_t idx = 0; idx < vec.size(); idx++) {
        // reset the vector, erase next value
        temp = vec;
        temp.erase(temp.begin() + idx);

        // check if safe
        if (checkSafe(temp)){
            safe++;
            break;
        }
    }
    return safe;
}
